### PR TITLE
Api tests and  System test PR

### PR DESCRIPTION
**Pull request by hrajput89**
#### Fixes #.

- [x] Ran `flake8`
- [x] Ran `run_repo_tests.sh`

#### Changes proposed in this pull request:

- API tests
- System Test

### CODE REVIEW

File: testsuites/CBLTester/System_test_multiple_device/conftest.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+        "sync_gateway_version": sync_gateway_version,
+        "target_admin_url": target_admin_url,
+        "enable_sample_bucket": enable_sample_bucket,
+        "cbl_db_list": cbl_db_list,
+        "db_name_list": db_name_list,
+        "base_url_list": base_url_list,
+        "query_obj_list": query_obj_list,
+        "sg_config": sg_config,
+        "db_obj_list": db_obj_list,
+        # "testserver_list": testserver_list,
```
Comment (raghusarangapani) [inline testsuites/CBLTester/System_test_multiple_device/conftest.py]: Is this needed?

Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/conftest.py]: removed it

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_document.py [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+        (random_string(6), "{}".format(random.uniform(1, 10))),
+        (random_string(6), "{}".format(random.uniform(11, 100))),
+        (random_string(6), "{}".format(random.uniform(101, 1000)))
+    ])
+    def test_get_set_double(self, key, value):
+        '''
+        @summary: Testing Get and Set Double method of Document API
+        '''
+        doc = self.doc_obj.create()
+        # double_obj = self.datatype.setDouble(value)
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_document.py]: done

File: testsuites/CBLTester/System_test_multiple_device/conftest.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+    parser.addoption("--cluster-config",
+                     action="store",
+                     help="Provide cluster config to use. Default is base config",
+                     default="base")
+
+
+# This will get called once before the first test that
+# runs with this as input parameters in this file
+# This setup will be called once for all tests in the
+# testsuites/CBLTester/CBL_Functional_tests/ directory
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/conftest.py]: changed the comments

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    complex_doc
+from datetime import datetime, timedelta
+
+
+@pytest.mark.sanity
+@pytest.mark.listener
+@pytest.mark.replication
+@pytest.mark.parametrize("num_of_docs, num_of_updates, num_of_docs_to_update, num_of_docs_in_itr, num_of_doc_to_delete, num_of_docs_to_add, up_time", [
+    (1000000, 100, 100, 1000, 1000, 2000, 4 * 24 * 60),
+    # (500, 5, 10, 5, 10, 10, 1 * 20),
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: No, this is just for quick check of the system. We run System test for larger data set. The 2nd argument is only for checking the state of environment for system test.

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+            t = Thread(target=_replicaton_status_check, args=(repl_obj, repl))
+            t.start()
+            t.join()
+            query.query_get_docs_limit_offset(cbl_db, limit=query_limit, offset=query_offset)
+            time.sleep(5)
+        _check_doc_count(db_obj_list, cbl_db_list)
+        # removing ids of deleted doc from the list
+        doc_ids = doc_ids - docs_to_delete
+        # Deleting a node from the cluster
+        print "Rebalance out server: {}".format(server.host)
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: done

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+                elif generator == "simple_user":
+                    data = simple_user()
+                else:
+                    data = simple()
+                data["channels"] = channels_sg
+                data["_id"] = doc_id
+                added_docs[doc_id] = data
+                new_doc_ids.append(doc_id)
+            doc_ids.update(new_doc_ids)
+            print "creating {} docs on {} - {}".format(len(docs_to_create),
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: done

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+        time.sleep(5)
+    _check_doc_count(db_obj_list, cbl_db_list)
+
+
+def _replicaton_status_check(repl_obj, replicator):
+        repl_obj.wait_until_replicator_idle(replicator)
+        total = repl_obj.getTotal(replicator)
+        completed = repl_obj.getCompleted(replicator)
+        assert total == completed, "total is not equal to completed"
+        time.sleep(5)  # wait until replication is over
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: removed

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+
+
+def _check_doc_count(db_obj_list, cbl_db_list):
+    new_docs_count = set([db_obj.getCount(cbl_db) for db_obj, cbl_db in zip(db_obj_list, cbl_db_list)])
+    print "Doc count is - {}".format(new_docs_count)
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: done

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+        #######################################
+        docs_to_update = random.sample(doc_ids, num_of_docs_to_update)
+        i = 0
+        for db_obj, cbl_db, repl_obj, repl, query in zip(db_obj_list,
+                                                         cbl_db_list,
+                                                         replicator_obj_list,
+                                                         replicator_list,
+                                                         query_obj_list):
+            updates_per_db = len(docs_to_update) / len(db_obj_list)
+            print "Updating {} docs on {} db - {}".format(updates_per_db,
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: done

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+            db_obj.update_bulk_docs(cbl_db, num_of_updates, list(docs_to_update)[i: i + updates_per_db])
+            i += updates_per_db
+            # updating docs will affect all dbs as they are synced with SG.
+            t = Thread(target=_replicaton_status_check, args=(repl_obj, repl))
+            t.start()
+            t.join()
+            query.query_get_docs_limit_offset(cbl_db, limit=query_limit, offset=query_offset)
+
+        ###########################
+        # Deleting doc on SG side #
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: yes, changed the string to docs

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+            query.query_get_docs_limit_offset(cbl_db, limit=query_limit, offset=query_offset)
+            time.sleep(5)
+        _check_doc_count(db_obj_list, cbl_db_list)
+        # removing ids of deleted doc from the list
+        doc_ids = doc_ids - docs_to_delete
+        # Deleting a node from the cluster
+        print "Rebalance out server: {}".format(server.host)
+        primary_server.rebalance_out(server_urls, server)
+        ############################
+        # Deleting doc on CBL side #
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: yes, changed to docs

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+    x = 1
+    while(running_time - current_time > timedelta(0)):
+
+        print '*' * 20
+        print "Starting iteration no. {} of system testing".format(x)
+        print '*' * 20
+        x += 1
+        server = servers[random.randint(0, len(servers) - 1)]
+        ######################################
+        # Checking for doc update on SG side #
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: done

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+        ######################################
+        docs_to_update = random.sample(doc_ids, num_of_docs_to_update)
+        sg_docs = sg_client.get_all_docs(url=sg_url, db=sg_db, auth=session)["rows"]
+        sg_docs = [doc for doc in sg_docs if doc["id"] in docs_to_update]
+        print "Updating {} docs on SG - {}".format(len(docs_to_update),
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: done

File: testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+    def test_exists(self, db_name):
+        '''
+        @summary: Testing exist method of Database API
+        '''
+        if len(db_name) >= 128 and (self.liteserv_platform != "ios" or self.liteserv_platform != "android"):
+            pytest.skip("Test not supported on .net platforms or the db name is longer than or equal to 128 characters")
+
+        db = self.db_obj.create(db_name)
+        path = self.db_obj.getPath(db)
+        print "path for db is ", path
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/CBL_Functional_tests/APITests/test_database.py]: done

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+    primary_server = cluster.servers[0]
+    servers = cluster.servers[1:]
+    if not resume_cluster:
+        # Reset cluster to ensure no data in system
+        cluster.reset(sg_config_path=sg_config)
+        sg_client.create_user(sg_admin_url, sg_db, username, password, channels=channels_sg)
+
+        # adding bulk docs to each db
+        for cbl_db, db_obj, db_name in zip(cbl_db_list, db_obj_list, db_name_list):
+            print "Adding doc on {} db".format(db_name)
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: done

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+        _check_doc_count(db_obj_list, cbl_db_list)
+        count = db_obj_list[0].getCount(cbl_db_list[0])
+        itr_count = count / query_limit
+        if itr_count == 0:
+            itr_count = 1
+        for _ in range(itr_count):
+            existing_docs = db_obj_list[0].getDocIds(cbl_db_list[0], query_limit, query_offset)
+            doc_ids.update(existing_docs)
+            query_offset += query_limit
+        print "{} Docs in DB".format(len(doc_ids))
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: done

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 17]
```diff
@@ -0,0 +1,299 @@
...
+    current_time = datetime.now()
+    running_time = current_time + timedelta(minutes=up_time)
+
+    _check_doc_count(db_obj_list, cbl_db_list)
+    x = 1
+    while(running_time - current_time > timedelta(0)):
+
+        print '*' * 20
+        print "Starting iteration no. {} of system testing".format(x)
+        print '*' * 20
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: done

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+            t.start()
+            t.join()
+            query.query_get_docs_limit_offset(cbl_db, limit=query_limit, offset=query_offset)
+
+        ###########################
+        # Deleting doc on SG side #
+        ###########################
+        docs_to_delete = set(random.sample(doc_ids, num_of_doc_to_delete))
+        sg_docs = sg_client.get_bulk_docs(url=sg_url, db=sg_db, doc_ids=list(docs_to_delete), auth=session)[0]
+        print "Deleting {} docs on SG - {}".format(len(docs_to_delete),
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: done

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 19]
```diff
@@ -0,0 +1,299 @@
...
+        ############################
+        docs_to_delete = set(random.sample(doc_ids, num_of_doc_to_delete))
+        docs_to_delete_per_db = len(docs_to_delete) / len(db_obj_list)
+        i = 0
+        for db_obj, cbl_db, repl_obj, repl, query in zip(db_obj_list,
+                                                         cbl_db_list,
+                                                         replicator_obj_list,
+                                                         replicator_list,
+                                                         query_obj_list):
+            print "deleting {} docs from {} db - {}".format(docs_to_delete_per_db,
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: done

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 20]
```diff
@@ -0,0 +1,299 @@
...
+                                                     query_obj_list):
+                t = Thread(target=_replicaton_status_check, args=(repl_obj, repl))
+                t.start()
+                t.join()
+                query.query_get_docs_limit_offset(cbl_db, limit=query_limit, offset=query_offset)
+        _check_doc_count(db_obj_list, cbl_db_list)
+        # removing ids of deleted doc from the list
+        doc_ids = doc_ids - docs_to_delete
+        # Adding the node back to the cluster
+        print "Adding Server back {}".format(server.host)
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: done

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 21]
```diff
@@ -0,0 +1,299 @@
...
+        primary_server.add_node(server, services="kv,index,n1ql")
+        print "Rebalance in server: {}".format(server.host)
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: done

File: testsuites/CBLTester/System_test_multiple_device/test_system_testing.py [hunk 22]
```diff
@@ -0,0 +1,299 @@
...
+            t.start()
+            t.join()
+            query.query_get_docs_limit_offset(cbl_db, limit=query_limit, offset=query_offset)
+            time.sleep(5)
+        doc_id_for_new_docs += num_of_docs_to_add
+        _check_doc_count(db_obj_list, cbl_db_list)
+
+        current_time = datetime.now()
+    # stopping replication
+    print "Test completed. Stopping Replicators"
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply testsuites/CBLTester/System_test_multiple_device/test_system_testing.py]: done

**Pull request merged**